include/openenclave/bits/device.h

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#ifndef _OE_BITS_DEVICE_H
+#define _OE_BITS_DEVICE_H
+
+#include <openenclave/bits/defs.h>
+#include <openenclave/bits/result.h>
+#include <openenclave/bits/types.h>
+
+OE_EXTERNC_BEGIN
+
+/**
+ * Device identifiers known by Open Enclave.
+ */
+enum
+{
+    /** The null device id. */
+    OE_DEVID_NONE,
+
+    /** The non-secure host file system. */
+    OE_DEVID_HOSTFS,
+
+    /** The Intel SGX protected file system. */
+    OE_DEVID_SGXFS,
+
+    /** The non-secure host socket device. */
+    OE_DEVID_HOSTSOCK,
+
+    /** The secure enclave sockets device (future). */
+    OE_DEVID_ENCLAVESOCK,
+
+    /** The host epoll device. */
+    OE_DEVID_HOSTEPOLL,
+
+    /** The host eventfd device. */
+    OE_DEVID_EVENTFD,
+};
+
+/**
+ * Associate a device id with the current thread.
+ *
+ * This function associates the given device id with the current thread.
+ * It sets this device id into a thread-local-storage slot, which affects
+ * the behaviour of the following libc functions:
+ *
+ *     - open()
+ *     - rename()
+ *     - access()
+ *     - truncate()
+ *     - link()
+ *     - unlink()
+ *     - rmdir()
+ *     - opendir()
+ *     - stat()
+ *     - mkdir()
+ *     - socket()
+ *
+ * By default, these functions use either the mounter (file I/O) or the
+ * address family (socket I/O) to select a device for the given operation.
+ * But this function bypasses these device resolution methods and uses the
```

Comment:
I moved this to internal for now. This is to support Dave's functions that resolve the device by the first parameter. These functions should have the following form:

```
int foo(uint64_t devide, …)
{
    oe_set_thread_device(...);

    /* perform I/O operation. */

    oe_clear_thread_devid();
}
```

The two functions must be paired, else the caller might get undefined results. I retracted this form the public API for now. When Dave uses this, we will figure out how it would be published.